File: Progress-sheet/0x04-Sorting/find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int k = p.size();
        vector<int> ans;

        vector<int> window(128);
        vector<int>  sample(128);
        for(int i = 0; i< p.size(); i++){
            sample[p[i]]++;
        }
        for (int i = 0,l=0; i < s.length();++i)
        {
            window[s[i]]++;
            if(i - l + 1 == k){
                if(window == sample) {
                    ans.push_back(l);
                }
                --window[s[l++]];
                
            }
        }
        //cba

        return ans;
    }
};

File: Progress-Sheet/0x05-SlidingWindow/leetcode/find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        k = len(p)
        ans = []
        window = [0]*128
        sample = [0]*128

        for i in p:
            sample[ord(i)] += 1
        
        right = 0
        left = 0
        while(right < len(s)):
            window[ord(s[right])] += 1
            if(right - left+ 1 == k):
                if(window == sample):
                    ans.append(left)
                window[ord(s[left])]-=1
                left+=1
            right+=1
        return ans
            
        

